src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterCallbackHandler.java

```diff
@@ -66,17 +82,64 @@ private CallbackResponse prepopulateData(CallbackParams callbackParams) {
 
     public CallbackResponse createAndPreview(CallbackParams callbackParams) {
         String authorisation = callbackParams.getParams().get(CallbackParams.Params.BEARER_TOKEN).toString();
-        return generalLetterService.createAndPreview(
-            callbackParams.getRequest().getCaseDetails(),
-            authorisation,
-            DRAFT_LETTER_DOC,
-            generalLetterTemplateId);
+        try {
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .data(ImmutableMap.of(
+                    DRAFT_LETTER_DOC,
+                    CCDDocument.builder().documentUrl(generalLetterService.createAndPreview(
+                        caseDetailsConverter.extractCCDCase(callbackParams.getRequest().getCaseDetails()),
+                        authorisation,
+                        generalLetterTemplateId)).build()
+                ))
+                .build();
+        } catch (DocumentGenerationFailedException e) {
+            logger.info("General Letter creating and preview failed", e);
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .errors(Collections.singletonList(ERROR_MESSAGE))
+                .build();
+        }
     }
 
     public CallbackResponse printAndUpdateCaseDocuments(CallbackParams callbackParams) {
+        CCDCase ccdCase = caseDetailsConverter.extractCCDCase(callbackParams.getRequest().getCaseDetails());
         String authorisation = callbackParams.getParams().get(CallbackParams.Params.BEARER_TOKEN).toString();
-        return generalLetterService.printAndUpdateCaseDocuments(
-            callbackParams.getRequest().getCaseDetails(),
-            authorisation);
+        boolean errors = false;
+        CCDCase updatedCcdCase = ccdCase;
+        try {
+            updatedCcdCase = generalLetterService.printAndUpdateCaseDocuments(
+                ccdCase,
+                caseDetailsConverter.extractClaim(callbackParams.getRequest().getCaseDetails()),
+                authorisation,
+                getDocumentName(ccdCase)
+            );
+        } catch (Exception e) {
+            logger.info("General Letter printing and case documents update failed", e);
+            errors = true;
+        }
+        if (!errors) {
+            logger.info("General Letter: updating case document with general letter");
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .data(caseDetailsConverter.convertToMap(updatedCcdCase))
+                .build();
+        } else {
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .errors(Collections.singletonList(ERROR_MESSAGE))
+                .build();
+        }
+    }
+
+    private String getDocumentName(CCDCase ccdCase) {
+        Integer number = Math.toIntExact(ccdCase.getCaseDocuments()
```

Comment:
Probably best to change this to a long rather than wrap it in this  (i didn't realise that count() returns a long)
```suggestion
        long number = ccdCase.getCaseDocuments()
```